cinder/backup/chunkeddriver.py

```diff
@@ -780,3 +776,161 @@ def delete_backup(self, backup):
                 eventlet.sleep(0)
 
         LOG.debug('delete %s finished.', backup['id'])
+
+
+class BackupRestoreHandle(object):
+    """Class used to reconstruct a backup from chunks."""
+    def __init__(self, chunked_driver):
+        self._driver = chunked_driver
+        self._segments = []
+        self._object_readers = {}
+        self._idx = -1
+
+    def has_next(self):
+        return self._idx < len(self._segments) - 1
+
+    def get_next(self):
+        self._idx += 1
+        return self._segments[self._idx]
+
+    def read(self, segment):
+        """Reads the bytes of a segment"""
+        buff_reader = self._get_reader(segment)
+        # seek inside the backup chunk containing this segment
+        offset_diff = segment.offset - segment.obj['offset']
+        buff_reader.seek(offset_diff)
+        # read the segment's length from the chunk
+        data = buff_reader.read(segment.length)
+        eventlet.tpool.execute(self._clear_reader, segment)
+        return data
+
+    def _get_reader(self, segment):
+        """Keeps an internal cache of object readers.
```

Comment:
Normally a sort of IO error is to be thrown, which is eventually failing the restore process.
This would be fine for the majority of drivers maybe, since their readers are not network connections.
What would you think? Adding a retrial mechanism that would attempt to obtain a new reader in case of an reading error, and fail after `x` trials?

Reply:
Yeah, that sounds reasonable.

Reply:
Actually, our cache is made of `BytesIO`. We read the chunk bytes from the storage backend and save it in `BytesIO` (useful to seek for the inner segments). Once we're done writing, we close and remove that `BytesIO`.
Therefore, we read the chunk from the storage backend in one go. I don't think we would get connection loss in this case, at least not because of us doing long running stuff.

Reply:
Oh, right. We only have ~50MB chunks, right? Then you're probably right and we wouldn't need it.

If we keep those 50 MB in memory for all segments of a chunk, doesn't that consume a lot of memory? There's no other way because of encryption and compression, I guess.

Reply:
It keeps the 50MB per chunk, so only once. The segments are then seeked from those 50MB.
Memory consumption grows with the number of parent backups, depending on how many overwrites needs to be done on a particular sector of the backup.
Another way would be to call the storage backend for each segment of a chunk and not keep this cache, which is not convenient because that would slow the process a lot.

Reply:
Additionally we could use the local filesystem for caching instead of the memory.
It would't require a lot of space.

Reply:
If we use a disk cache, then on failed backups would we leave behind segments on disk?